Builder/AbstractFormContractor.php

```diff
@@ -53,97 +56,127 @@ public function __construct(FormFactoryInterface $formFactory)
-            $fieldDescription->setOption('edit', $fieldDescription->getOption('edit', 'admin'));
-        } elseif ($type === 'sonata_type_collection') {
-            if (!$fieldDescription->describesCollectionValuedAssociation()) {
-                throw new \LogicException(sprintf(
-                    'The `sonata_type_collection` type only handles collection-valued associations.'
-                    .' Try using `sonata_type_admin` or one of the `sonata_type_model` variants'
-                    .' for field `%s`',
-                    $fieldDescription->getName()
-                ));
-            }
-            $options['type'] = 'sonata_type_admin';
-            $options['modifiable'] = true;
-            $options['type_options'] = array(
-                'sonata_field_description' => $fieldDescription,
-                'data_class' => $fieldDescription->getAssociationAdmin()->getClass(),
-            );
+        switch ($fqcn) {
+            case 'Sonata\AdminBundle\Form\Type\ModelType':
+            case 'Sonata\AdminBundle\Form\Type\ModelAutocompleteType':
+                if ($fieldDescription->describesCollectionValuedAssociation()) {
+                    $options['multiple'] = true;
+                }
+                $options['class'] = $fieldDescription->getTargetEntity();
+                $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
+                break;
+            case 'Sonata\AdminBundle\Form\Type\ModelTypeList':
+            case 'Sonata\AdminBundle\Form\Type\ModelHiddenType':
+            case 'Sonata\AdminBundle\Form\Type\ModelReferenceType':
+                $options['class'] = $fieldDescription->getTargetEntity();
+                $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
+                break;
+            case 'Sonata\AdminBundle\Form\Type\AdminType':
+                $options['delete'] = false;
+                $options['data_class'] = $fieldDescription->getTargetEntity();
+                break;
+            case 'Sonata\CoreBundle\Form\Type\CollectionType':
+                // NEXT_MAJOR: replace by FQCN when dropping Symfony <2.8 support
+                $options['type'] = 'sonata_type_admin';
+                $options['modifiable'] = true;
+                $options['type_options'] = array(
+                    'sonata_field_description' => $fieldDescription,
+                    'data_class' => $fieldDescription->getTargetEntity(),
+                );
+                break;
         }
 
         return $options;
     }
 
     /**
+     * Ensures that the given field description has an associated target entity.
+     *
+     * @param FieldDescriptionInterface $fieldDescription
+     *
+     * @throws MissingTargetEntityException if the field description has no target entity
+     */
+    final protected function ensureTargetEntity(FieldDescriptionInterface $fieldDescription)
```

Comment:
I made it `final protected` to be used in subclasses.
Because persistence bundles are allowed to add specific widget types (like the PHPCR tree widget for parent/child relationships) and might want to use these generic validation methods.